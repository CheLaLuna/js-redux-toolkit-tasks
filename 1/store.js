import omit from "lodash/omit";
import { legacy_createStore as createStore } from "redux";

// BEGIN (write your solution here)

// END
