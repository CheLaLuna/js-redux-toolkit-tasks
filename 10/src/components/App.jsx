import React from 'react';

import TodoBox from './TodoBox.jsx';

const App = () => (
  <div className="col-5">
    <TodoBox />
  </div>
);

export default App;
