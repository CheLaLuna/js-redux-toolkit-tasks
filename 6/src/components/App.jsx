import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import axios from 'axios';
import Posts from './Posts.jsx';
import routes from '../routes.js';

import { actions as usersActions } from '../slices/usersSlice.js';
import { actions as postsActions } from '../slices/postsSlice.js';
import { actions as commentsActions } from '../slices/commentsSlice.js';

import getNormalized from '../utilities/getNormalized.js';

const App = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    const fetchData = async () => {
      const { data } = await axios.get(routes.getData());
      const normalizedData = getNormalized(data);
      const {
        users,
        posts,
        comments,
      } = normalizedData.entities;

      dispatch(usersActions.setUsers({ entities: users, ids: Object.keys(users) }));
      dispatch(postsActions.setPosts({ entities: posts, ids: Object.keys(posts) }));
      dispatch(commentsActions.setComments({ entities: comments, ids: Object.keys(comments) }));
    };

    fetchData();
  });

  return (
    <div className="col-5">
      <Posts />
    </div>
  );
};

export default App;
