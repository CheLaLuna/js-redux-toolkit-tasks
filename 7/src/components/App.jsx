import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import axios from 'axios';
import Posts from './Posts.jsx';
import routes from '../routes.js';
import PostForm from './PostForm.jsx';

import { actions as usersActions } from '../slices/usersSlice.js';
import { actions as postsActions } from '../slices/postsSlice.js';
import { actions as commentsActions } from '../slices/commentsSlice.js';

import getNormalized from '../utilities/getNormalized.js';

const App = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    const fetchData = async () => {
      const { data } = await axios.get(routes.getData());
      const normalizedData = getNormalized(data);
      const {
        users,
        posts,
        comments,
      } = normalizedData.entities;

      dispatch(usersActions.addUsers(users));
      dispatch(postsActions.addPosts(posts));
      dispatch(commentsActions.addComments(comments));
    };

    fetchData();
  });

  return (
    <>
      <div className="card">
        <div className="card-header">
          Создать пост
        </div>
        <div className="card-body">
          <PostForm />
        </div>
      </div>
      <div className="mt-5">
        <Posts />
      </div>
    </>
  );
};

export default App;
