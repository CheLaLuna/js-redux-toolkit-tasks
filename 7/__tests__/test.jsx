import '@testing-library/jest-dom';

import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import nock from 'nock';
import App from '../src/components/App.jsx';
import store from '../src/slices/index.js';

test('Work 1', async () => {
  const blogPosts = [
    {
      id: 'post1',
      author: { id: 'user1', username: 'user1', name: 'User 1' },
      body: 'Первый пост',
      comments: [
        {
          id: 'comment1',
          author: { id: 'user2', username: 'user2', name: 'User 2' },
          text: 'Первый комментарий',
        },
        {
          id: 'comment2',
          author: { id: 'user3', username: 'user3', name: 'User 3' },
          text: 'Второй комментарий',
        },
      ],
    },
    {
      id: 'post2',
      author: { id: 'user2', username: 'user2', name: 'User 2' },
      body: 'Второй пост',
      comments: [],
    },
  ];
  nock('http://localhost').get('/api/data').reply(200, blogPosts);

  const vdom = (
    <Provider store={store}>
      <App />
    </Provider>
  );

  const { asFragment } = render(vdom);
  expect(await screen.findByText(`${blogPosts[0].body} - ${blogPosts[0].author.name}`)).toBeInTheDocument();
  expect(await screen.findByText(`${blogPosts[1].body} - ${blogPosts[1].author.name}`)).toBeInTheDocument();

  expect(await screen.findAllByText(blogPosts[0].author.name)).toHaveLength(3);
  expect(await screen.findByText(blogPosts[0].comments[0].text)).toBeInTheDocument();
  expect(await screen.findAllByText(blogPosts[0].comments[0].author.name)).toHaveLength(4);
  expect(await screen.findByText(blogPosts[0].comments[1].text)).toBeInTheDocument();
  expect(await screen.findAllByText(blogPosts[0].comments[1].author.name)).toHaveLength(4);

  await userEvent.type(screen.getByLabelText(/текст/i), 'Третий пост');
  await userEvent.selectOptions(screen.getByLabelText('Автор поста'), [blogPosts[0].author.name]);
  await userEvent.click(screen.getByRole('button', { name: /создать пост/i }));
  expect(await screen.findByText(`Третий пост - ${blogPosts[0].author.name}`)).toBeInTheDocument();
  expect(await screen.findAllByText(blogPosts[0].author.name)).toHaveLength(4);

  await userEvent.type(screen.getAllByLabelText(/комментарий/i)[0], 'Комментарий к 3 посту');
  await userEvent.selectOptions(screen.getAllByLabelText('Автор комментария')[0], [blogPosts[0].author.name]);
  await userEvent.click(screen.getAllByRole('button', { name: /создать комментарий/i })[0]);

  expect(screen.getByText('Комментарий к 3 посту')).toBeInTheDocument();
  expect(await screen.findAllByText(blogPosts[0].author.name)).toHaveLength(5);

  expect(asFragment()).toMatchSnapshot();
});
