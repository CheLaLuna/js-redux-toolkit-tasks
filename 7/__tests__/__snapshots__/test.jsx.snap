// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Work 1 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Создать пост
    </div>
    <div
      class="card-body"
    >
      <div
        class="m-3"
      >
        <form
          class=""
        >
          <div
            class="row"
          >
            <label
              class="form-label"
              for="body"
            >
              Текст
            </label>
            <textarea
              class="form-control"
              id="body"
              name="body"
              rows="3"
            />
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
              for="postAuthor"
            >
              Автор поста
            </label>
            <select
              class="form-control"
              id="postAuthor"
              name="author"
            >
              <option
                value=""
              />
              <option
                value="user1"
              >
                User 1
              </option>
              <option
                value="user2"
              >
                User 2
              </option>
              <option
                value="user3"
              >
                User 3
              </option>
            </select>
          </div>
          <div
            class="mt-3 row"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Создать пост
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    class="mt-5"
  >
    <div
      class="mt-3"
    >
      <div
        class="card mb-5"
      >
        <div
          class="card-header"
        >
          Первый пост - User 1
        </div>
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            User 2
          </h5>
          <p
            class="card-text"
          >
            Первый комментарий
          </p>
          <h5
            class="card-title"
          >
            User 3
          </h5>
          <p
            class="card-text"
          >
            Второй комментарий
          </p>
          <h5
            class="card-title"
          >
            User 1
          </h5>
          <p
            class="card-text"
          >
            Комментарий к 3 посту
          </p>
          <hr />
          <div
            class="m-3"
          >
            <form
              class=""
            >
              <div
                class="row"
              >
                <label
                  class="form-label"
                  for="text"
                >
                  Комментарий
                </label>
                <textarea
                  class="form-control"
                  id="text"
                  name="text"
                  rows="3"
                />
              </div>
              <div
                class="row"
              >
                <label
                  class="form-label col-form-label col-sm-2"
                  for="author"
                >
                  Автор комментария
                </label>
                <select
                  class="form-control"
                  id="author"
                  name="author"
                >
                  <option
                    value=""
                  />
                  <option
                    value="user1"
                  >
                    User 1
                  </option>
                  <option
                    value="user2"
                  >
                    User 2
                  </option>
                  <option
                    value="user3"
                  >
                    User 3
                  </option>
                </select>
              </div>
              <div
                class="mt-3 row"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Создать комментарий
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        class="card mb-5"
      >
        <div
          class="card-header"
        >
          Второй пост - User 2
        </div>
        <div
          class="card-body"
        >
          <hr />
          <div
            class="m-3"
          >
            <form
              class=""
            >
              <div
                class="row"
              >
                <label
                  class="form-label"
                  for="text"
                >
                  Комментарий
                </label>
                <textarea
                  class="form-control"
                  id="text"
                  name="text"
                  rows="3"
                />
              </div>
              <div
                class="row"
              >
                <label
                  class="form-label col-form-label col-sm-2"
                  for="author"
                >
                  Автор комментария
                </label>
                <select
                  class="form-control"
                  id="author"
                  name="author"
                >
                  <option
                    value=""
                  />
                  <option
                    value="user1"
                  >
                    User 1
                  </option>
                  <option
                    value="user2"
                  >
                    User 2
                  </option>
                  <option
                    value="user3"
                  >
                    User 3
                  </option>
                </select>
              </div>
              <div
                class="mt-3 row"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Создать комментарий
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        class="card mb-5"
      >
        <div
          class="card-header"
        >
          Третий пост - User 1
        </div>
        <div
          class="card-body"
        >
          <hr />
          <div
            class="m-3"
          >
            <form
              class=""
            >
              <div
                class="row"
              >
                <label
                  class="form-label"
                  for="text"
                >
                  Комментарий
                </label>
                <textarea
                  class="form-control"
                  id="text"
                  name="text"
                  rows="3"
                />
              </div>
              <div
                class="row"
              >
                <label
                  class="form-label col-form-label col-sm-2"
                  for="author"
                >
                  Автор комментария
                </label>
                <select
                  class="form-control"
                  id="author"
                  name="author"
                >
                  <option
                    value=""
                  />
                  <option
                    value="user1"
                  >
                    User 1
                  </option>
                  <option
                    value="user2"
                  >
                    User 2
                  </option>
                  <option
                    value="user3"
                  >
                    User 3
                  </option>
                </select>
              </div>
              <div
                class="mt-3 row"
              >
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Создать комментарий
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
